Cabal/src/Distribution/Simple/Glob.hs

```diff
@@ -283,73 +493,153 @@ matchDirFileGlobWithDie verbosity rip version dir filepath = case parseFileGlob
 -- The 'FilePath' argument is the directory that the glob is relative
 -- to. It must be a valid directory (and hence it can't be the empty
 -- string). The returned values will not include this prefix.
-runDirFileGlob :: Verbosity -> FilePath -> Glob -> IO [GlobResult FilePath]
-runDirFileGlob verbosity rawDir pat = do
+runDirFileGlob
+  :: Verbosity
+  -> Maybe CabalSpecVersion
+  -- ^ If the glob we are running should care about the cabal spec, and warnings such as 'GlobWarnMultiDot', then this should be the version.
+  -- If you want to run a glob but don't care about any of the cabal-spec restrictions on globs, use 'Nothing'!
+  -> FilePath
+  -> FilePathGlobRel
+  -> IO [GlobResult FilePath]
+runDirFileGlob verbosity mspec rawRoot pat = do
```

Comment:
It is a big function, would it be hard to write unit tests?

Reply:
Thanks this makes it clearer (perhaps consider adding this as a comment). What you just said suggests changing `Glob` to represent the semantic of the globbing rather than the only the syntax (I assume they diverge in few specific cases); and adding a normalization function `normaliseForCabalSpec :: CabalSpecVersion -> Glob -> Glob`? (maybe even distinguishing the two `Glob` at the type level). Then`runDirFileGlob` is a mere interpreted independent from the context.

I am looking again and I think the code would be more readable if done in this way. E.g. ```
  isFile <- maybe (return True) (const $ doesFileExist (root </> dir </> s)) mspec
```
is a bit of a lie right? it actually mean `isFileOrMSpecIsNull`.

I really think there is a better way to write this.